cli/explain/compactgraph/compactgraph.go

```diff
@@ -0,0 +1,506 @@
+		diffWG.Add(1)
+		go func() {
+			aSpawn := old[output]
+			bSpawn := new[output]
+			spawnDiff, localChange, affectedBy := diffSpawns(aSpawn, bSpawn)
+			diffResults.Store(output, &diffResult{
+				spawnDiff:   spawnDiff,
+				localChange: localChange,
+				affectedBy:  affectedBy,
+			})
+			diffWG.Done()
+		}()
+	}
+	diffWG.Wait()
+
+	// Sort the common outputs topologically according to the new graph structure.
+	newOutputsSorted := new.sortedPrimaryOutputs()
+	commonOutputsSet := make(map[string]struct{})
+	for _, output := range commonOutputs {
+		commonOutputsSet[output] = struct{}{}
+	}
+	var commonOutputsSorted []string
+	for _, x := range newOutputsSorted {
+		if _, ok := commonOutputsSet[x]; ok {
+			commonOutputsSorted = append(commonOutputsSorted, x)
+		}
+	}
+
+	// Visit spawns in topological order and attribute their diffs to transitive dependencies if possible.
+	for _, output := range commonOutputsSorted {
+		resultEntry, _ := diffResults.Load(output)
+		result := resultEntry.(*diffResult)
+		spawn := new[output]
+		foundTransitiveCause := false
+		for _, affectedBy := range result.affectedBy {
+			if otherResultEntry, ok := diffResults.Load(affectedBy); ok {
+				foundTransitiveCause = true
+				otherDiff := otherResultEntry.(*diffResult).spawnDiff
+				if otherDiff.TransitivelyInvalidated == nil {
+					otherDiff.TransitivelyInvalidated = make(map[string]uint32)
+				}
+				for k, v := range result.spawnDiff.TransitivelyInvalidated {
+					otherDiff.TransitivelyInvalidated[k] += v
+				}
+				otherDiff.TransitivelyInvalidated[spawn.Mnemonic]++
+			}
+		}
+		if len(result.spawnDiff.Diffs) > 0 && (result.localChange || !foundTransitiveCause) {
+			spawnDiffs = append(spawnDiffs, result.spawnDiff)
+		}
+	}
+
+	return spawnDiffs
+}
+
+// reduceToRoots returns the sorted subset of outputs that are not inputs to any other spawn in the graph.
+func (cg *CompactGraph) reduceToRoots(outputs []string) []string {
+	rootsSet := make(map[string]struct{})
+	for _, output := range outputs {
+		rootsSet[output] = struct{}{}
+	}
+
+	// Visit all nodes in the graph and remove those with incoming edges from the set of roots.
+	var toVisit []interface{}
+	visited := make(map[interface{}]struct{})
+	markForVisit := func(n interface{}) {
+		if _, seen := visited[n]; !seen {
+			toVisit = append(toVisit, n)
+		}
+	}
+	for _, root := range outputs {
+		markForVisit((*cg)[root])
+	}
+	for len(toVisit) > 0 {
+		var node interface{}
+		node, toVisit = toVisit[0], toVisit[1:]
+		visited[node] = struct{}{}
+		switch n := node.(type) {
+		case *File:
+		case *Directory:
+			delete(rootsSet, n.Path())
+		}
+		cg.visitSuccessors(node, func(input interface{}) {
+			markForVisit(input)
+		})
+	}
+
+	roots := maps.Keys(rootsSet)
+	slices.Sort(roots)
+	return roots
+}
+
+// sortedPrimaryOutputs returns the primary output paths of the spawns in topological order.
+func (cg *CompactGraph) sortedPrimaryOutputs() []string {
+	var toVisit []interface{}
```

Comment:
nit: this could be `[]*Spawn` no?

Reply:
Ah, no, because the nodes in the graph are not just spawns, but also any kind of `Input`.